### PR TITLE
[CELEBORN-1477][FOLLOWUP] Shade the dependencies for openapi-client

### PR DESCRIPTION
**Pull request by turboFei**
### What changes were proposed in this pull request?

Shade the dependencies for openapi-client.

### Why are the changes needed?

To prevent dependency conflicts when the openapi-client involved in client.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass the UT.

### CODE REVIEW

**Comment by turboFei** [comment]
cc @pan3793

**Comment by pan3793** [comment]
sorry, my mbp is damaged and I don't have a full dev env now, could you please manually check the final jar and paste the `tree`(just some top levels) output to ensure no class leak?

**Comment by turboFei** [comment]
> BTW, what's the sbt side?

Checking.

**Comment by turboFei** [comment]
For mvn, 
```
(base) ➜  celeborn git:(openapi_client_shade) jar -tf openapi/openapi-client/target/celeborn-openapi-client_2.12-0.6.0-SNAPSHOT.jar|grep -v celeborn
META-INF/
META-INF/MANIFEST.MF
org/
org/apache/
META-INF/NOTICE
META-INF/maven/
META-INF/DEPENDENCIES
META-INF/LICENSE
META-INF/maven/io.swagger/
META-INF/maven/io.swagger/swagger-annotations/
META-INF/maven/io.swagger/swagger-annotations/pom.xml
META-INF/maven/io.swagger/swagger-annotations/pom.properties
META-INF/services/
META-INF/maven/org.openapitools/
META-INF/maven/org.openapitools/jackson-databind-nullable/
META-INF/maven/org.openapitools/jackson-databind-nullable/pom.xml
META-INF/maven/org.openapitools/jackson-databind-nullable/pom.properties
META-INF/LICENSE.md
META-INF/NOTICE.md
META-INF/maven/org.glassfish.jersey.core/
META-INF/maven/org.glassfish.jersey.core/jersey-client/
META-INF/maven/org.glassfish.jersey.core/jersey-client/pom.properties
META-INF/maven/org.glassfish.jersey.core/jersey-client/pom.xml
META-INF/native-image/
META-INF/native-image/org.glassfish.jersey.core/
META-INF/native-image/org.glassfish.jersey.core/jersey-client/
META-INF/native-image/org.glassfish.jersey.core/jersey-client/reflect-config.json
META-INF/native-image/org.glassfish.jersey.core/jersey-client/resource-config.json
META-INF/NOTICE.markdown
META-INF/maven/org.glassfish.jersey.media/
META-INF/maven/org.glassfish.jersey.media/jersey-media-json-jackson/
META-INF/maven/org.glassfish.jersey.media/jersey-media-json-jackson/pom.properties
META-INF/maven/org.glassfish.jersey.media/jersey-media-json-jackson/pom.xml
META-INF/maven/com.fasterxml.jackson.core/
META-INF/maven/com.fasterxml.jackson.core/jackson-annotations/
META-INF/maven/com.fasterxml.jackson.core/jackson-annotations/pom.xml
META-INF/maven/com.fasterxml.jackson.core/jackson-annotations/pom.properties
META-INF/maven/com.fasterxml.jackson.core/jackson-databind/
META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.xml
META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.properties
META-INF/versions/
META-INF/versions/9/
META-INF/versions/9/module-info.class
META-INF/maven/com.fasterxml.jackson.datatype/
META-INF/maven/com.fasterxml.jackson.datatype/jackson-datatype-jsr310/
META-INF/maven/com.fasterxml.jackson.datatype/jackson-datatype-jsr310/pom.properties
META-INF/maven/com.fasterxml.jackson.datatype/jackson-datatype-jsr310/pom.xml
META-INF/maven/org.glassfish.jersey.media/jersey-media-multipart/
META-INF/maven/org.glassfish.jersey.media/jersey-media-multipart/pom.properties
META-INF/maven/org.glassfish.jersey.media/jersey-media-multipart/pom.xml
META-INF/services/jakarta.validation.valueextraction.ValueExtractor
META-INF/services/javax.validation.valueextraction.ValueExtractor
(base) ➜  celeborn git:(openapi_client_shade) 

```

```
(base) ➜  celeborn git:(openapi_client_shade) jar -tf openapi/openapi-client/target/celeborn-openapi-client_2.12-0.6.0-SNAPSHOT.jar|grep  'celeborn/shaded'|wc -l
    1363

```

**Comment by pan3793** [comment]
```
META-INF/services/jakarta.validation.valueextraction.ValueExtractor
META-INF/services/javax.validation.valueextraction.ValueExtractor
```

are they compile/runtime deps? why ServicesResourceTransformer does not cover them?

**Comment by turboFei** [comment]
For sbt:
```
(base) ➜  celeborn git:(openapi_client_shade) ✗ jar -tf openapi/openapi-client/target/scala-2.12/celeborn-openapi-client_2.12-0.6.0-SNAPSHOT.jar|grep -v celeborn            
META-INF/
META-INF/MANIFEST.MF
META-INF/maven/
META-INF/maven/com.fasterxml.jackson.core/
META-INF/maven/com.fasterxml.jackson.core/jackson-annotations/
META-INF/maven/com.fasterxml.jackson.core/jackson-annotations/pom.properties
META-INF/maven/com.fasterxml.jackson.core/jackson-annotations/pom.xml
META-INF/maven/com.fasterxml.jackson.core/jackson-databind/
META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.properties
META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.xml
META-INF/maven/com.fasterxml.jackson.datatype/
META-INF/maven/com.fasterxml.jackson.datatype/jackson-datatype-jsr310/
META-INF/maven/com.fasterxml.jackson.datatype/jackson-datatype-jsr310/pom.properties
META-INF/maven/com.fasterxml.jackson.datatype/jackson-datatype-jsr310/pom.xml
META-INF/maven/io.swagger/
META-INF/maven/io.swagger/swagger-annotations/
META-INF/maven/io.swagger/swagger-annotations/pom.properties
META-INF/maven/io.swagger/swagger-annotations/pom.xml
META-INF/maven/io.swagger/swagger-models/
META-INF/maven/io.swagger/swagger-models/pom.properties
META-INF/maven/io.swagger/swagger-models/pom.xml
META-INF/maven/org.glassfish.jersey.core/
META-INF/maven/org.glassfish.jersey.core/jersey-client/
META-INF/maven/org.glassfish.jersey.core/jersey-client/pom.properties
META-INF/maven/org.glassfish.jersey.core/jersey-client/pom.xml
META-INF/maven/org.glassfish.jersey.media/
META-INF/maven/org.glassfish.jersey.media/jersey-media-json-jackson/
META-INF/maven/org.glassfish.jersey.media/jersey-media-json-jackson/pom.properties
META-INF/maven/org.glassfish.jersey.media/jersey-media-json-jackson/pom.xml
META-INF/maven/org.glassfish.jersey.media/jersey-media-multipart/
META-INF/maven/org.glassfish.jersey.media/jersey-media-multipart/pom.properties
META-INF/maven/org.glassfish.jersey.media/jersey-media-multipart/pom.xml
META-INF/maven/org.openapitools/
META-INF/maven/org.openapitools/jackson-databind-nullable/
META-INF/maven/org.openapitools/jackson-databind-nullable/pom.properties
META-INF/maven/org.openapitools/jackson-databind-nullable/pom.xml
META-INF/native-image/
META-INF/native-image/org.glassfish.jersey.core/
META-INF/native-image/org.glassfish.jersey.core/jersey-client/
META-INF/native-image/org.glassfish.jersey.core/jersey-client/reflect-config.json
META-INF/native-image/org.glassfish.jersey.core/jersey-client/resource-config.json
META-INF/services/
META-INF/services/jakarta.validation.valueextraction.ValueExtractor
META-INF/services/javax.validation.valueextraction.ValueExtractor
META-INF/versions/
META-INF/versions/9/
META-INF/versions/9/module-info.class
module-info.class
org/
org/apache/
(base) ➜  celeborn git:(openapi_client_shade) 

```


```
jar -tf openapi/openapi-client/target/scala-2.12/celeborn-openapi-client_2.12-0.6.0-SNAPSHOT.jar|grep 'celeborn/shaded'|wc -l
    1468
```

**Comment by turboFei** [comment]
> META-INF/services/jakarta.validation.valueextraction.ValueExtractor
>   META-INF/services/javax.validation.valueextraction.ValueExtractor 

Looks like interface

**Comment by turboFei** [comment]
After add swagger-models for mvn.

For the jar build with mvn:
```
(base) ➜  celeborn git:(openapi_client_shade) jar -tf openapi/openapi-client/target/celeborn-openapi-client_2.12-0.6.0-SNAPSHOT.jar|grep 'celeborn/shaded'|wc -l
    1468
```

For the jar build with sbt:


```
(base) ➜  celeborn git:(openapi_client_shade) jar -tf openapi/openapi-client/target/scala-2.12/celeborn-openapi-client_2.12-0.6.0-SNAPSHOT.jar|grep 'celeborn/shaded' |wc -l
    1468
(base) ➜  celeborn git:(openapi_client_shade) 
```

File: project/CelebornBuild.scala [hunk 1]
```diff
@@ -1322,7 +1323,53 @@ object CelebornOpenApi {
...
+      (assembly / assemblyPackageScala / assembleArtifact) := false,
+
+      (assembly / assemblyExcludedJars) := {
+        val cp = (assembly / fullClasspath).value
+        cp filter { v =>
+          val name = v.data.getName
+          !(name.startsWith("celeborn-") ||
+              name.startsWith("swagger-annotations-") ||
+              name.startsWith("swagger-models-") ||
+              name.startsWith("jackson-databind-") ||
```
Comment (turboFei) [inline project/CelebornBuild.scala:1346]: <include>org.openapitools:jackson-databind-nullable</include>
              <include>com.fasterxml.jackson.core:jackson-databind</include>

File: openapi/openapi-client/pom.xml [hunk 2]
```diff
@@ -197,6 +201,70 @@
...
+                <exclude>**/log4j.properties</exclude>
+                <exclude>META-INF/LICENSE.txt</exclude>
+                <exclude>META-INF/NOTICE.txt</exclude>
+                <exclude>LICENSE.txt</exclude>
+                <exclude>NOTICE.txt</exclude>
+              </excludes>
+            </filter>
+          </filters>
+          <transformers>
+            <transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer"></transformer>
```
Comment (cfmcgrady) [inline openapi/openapi-client/pom.xml:264]: Does the SBT also perform the same service resource transformation?

Comment (turboFei) [reply openapi/openapi-client/pom.xml:264]: Per the result, they are consistent:

For mvn:
```
(base) ➜  celeborn git:(openapi_client_shade) jar -tf openapi/openapi-client/target/celeborn-openapi-client_2.12-0.6.0-SNAPSHOT.jar|grep -v celeborn
META-INF/
META-INF/MANIFEST.MF
org/
org/apache/
META-INF/NOTICE
META-INF/maven/
META-INF/DEPENDENCIES
META-INF/LICENSE
META-INF/maven/io.swagger/
META-INF/maven/io.swagger/swagger-annotations/
META-INF/maven/io.swagger/swagger-annotations/pom.xml
META-INF/maven/io.swagger/swagger-annotations/pom.properties
META-INF/services/
META-INF/maven/org.openapitools/
META-INF/maven/org.openapitools/jackson-databind-nullable/
META-INF/maven/org.openapitools/jackson-databind-nullable/pom.xml
META-INF/maven/org.openapitools/jackson-databind-nullable/pom.properties
META-INF/LICENSE.md
META-INF/NOTICE.md
META-INF/maven/org.glassfish.jersey.core/
META-INF/maven/org.glassfish.jersey.core/jersey-client/
META-INF/maven/org.glassfish.jersey.core/jersey-client/pom.properties
META-INF/maven/org.glassfish.jersey.core/jersey-client/pom.xml
META-INF/native-image/
META-INF/native-image/org.glassfish.jersey.core/
META-INF/native-image/org.glassfish.jersey.core/jersey-client/
META-INF/native-image/org.glassfish.jersey.core/jersey-client/reflect-config.json
META-INF/native-image/org.glassfish.jersey.core/jersey-client/resource-config.json
META-INF/NOTICE.markdown
META-INF/maven/org.glassfish.jersey.media/
META-INF/maven/org.glassfish.jersey.media/jersey-media-json-jackson/
META-INF/maven/org.glassfish.jersey.media/jersey-media-json-jackson/pom.properties
META-INF/maven/org.glassfish.jersey.media/jersey-media-json-jackson/pom.xml
META-INF/maven/com.fasterxml.jackson.core/
META-INF/maven/com.fasterxml.jackson.core/jackson-annotations/
META-INF/maven/com.fasterxml.jackson.core/jackson-annotations/pom.xml
META-INF/maven/com.fasterxml.jackson.core/jackson-annotations/pom.properties
META-INF/maven/com.fasterxml.jackson.core/jackson-databind/
META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.xml
META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.properties
META-INF/versions/
META-INF/versions/9/
META-INF/versions/9/module-info.class
META-INF/maven/com.fasterxml.jackson.datatype/
META-INF/maven/com.fasterxml.jackson.datatype/jackson-datatype-jsr310/
META-INF/maven/com.fasterxml.jackson.datatype/jackson-datatype-jsr310/pom.properties
META-INF/maven/com.fasterxml.jackson.datatype/jackson-datatype-jsr310/pom.xml
META-INF/maven/org.glassfish.jersey.media/jersey-media-multipart/
META-INF/maven/org.glassfish.jersey.media/jersey-media-multipart/pom.properties
META-INF/maven/org.glassfish.jersey.media/jersey-media-multipart/pom.xml
META-INF/services/jakarta.validation.valueextraction.ValueExtractor
META-INF/services/javax.validation.valueextraction.ValueExtractor
(base) ➜  celeborn git:(openapi_client_shade) 

```

For sbt:
```
META-INF/
META-INF/MANIFEST.MF
META-INF/maven/
META-INF/maven/com.fasterxml.jackson.core/
META-INF/maven/com.fasterxml.jackson.core/jackson-annotations/
META-INF/maven/com.fasterxml.jackson.core/jackson-annotations/pom.properties
META-INF/maven/com.fasterxml.jackson.core/jackson-annotations/pom.xml
META-INF/maven/com.fasterxml.jackson.core/jackson-databind/
META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.properties
META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.xml
META-INF/maven/com.fasterxml.jackson.datatype/
META-INF/maven/com.fasterxml.jackson.datatype/jackson-datatype-jsr310/
META-INF/maven/com.fasterxml.jackson.datatype/jackson-datatype-jsr310/pom.properties
META-INF/maven/com.fasterxml.jackson.datatype/jackson-datatype-jsr310/pom.xml
META-INF/maven/io.swagger/
META-INF/maven/io.swagger/swagger-annotations/
META-INF/maven/io.swagger/swagger-annotations/pom.properties
META-INF/maven/io.swagger/swagger-annotations/pom.xml
META-INF/maven/io.swagger/swagger-models/
META-INF/maven/io.swagger/swagger-models/pom.properties
META-INF/maven/io.swagger/swagger-models/pom.xml
META-INF/maven/org.glassfish.jersey.core/
META-INF/maven/org.glassfish.jersey.core/jersey-client/
META-INF/maven/org.glassfish.jersey.core/jersey-client/pom.properties
META-INF/maven/org.glassfish.jersey.core/jersey-client/pom.xml
META-INF/maven/org.glassfish.jersey.media/
META-INF/maven/org.glassfish.jersey.media/jersey-media-json-jackson/
META-INF/maven/org.glassfish.jersey.media/jersey-media-json-jackson/pom.properties
META-INF/maven/org.glassfish.jersey.media/jersey-media-json-jackson/pom.xml
META-INF/maven/org.glassfish.jersey.media/jersey-media-multipart/
META-INF/maven/org.glassfish.jersey.media/jersey-media-multipart/pom.properties
META-INF/maven/org.glassfish.jersey.media/jersey-media-multipart/pom.xml
META-INF/maven/org.openapitools/
META-INF/maven/org.openapitools/jackson-databind-nullable/
META-INF/maven/org.openapitools/jackson-databind-nullable/pom.properties
META-INF/maven/org.openapitools/jackson-databind-nullable/pom.xml
META-INF/native-image/
META-INF/native-image/org.glassfish.jersey.core/
META-INF/native-image/org.glassfish.jersey.core/jersey-client/
META-INF/native-image/org.glassfish.jersey.core/jersey-client/reflect-config.json
META-INF/native-image/org.glassfish.jersey.core/jersey-client/resource-config.json
META-INF/services/
META-INF/services/jakarta.validation.valueextraction.ValueExtractor
META-INF/services/javax.validation.valueextraction.ValueExtractor
META-INF/versions/
META-INF/versions/9/
META-INF/versions/9/module-info.class
module-info.class
org/
org/apache/
```

Comment (turboFei) [reply openapi/openapi-client/pom.xml:264]: The shade mvn/sbt plugin setting are the same with the existing shade modules.

**Review by cfmcgrady — APPROVED** [review]

**Comment by turboFei** [comment]
rebased the code, cc @pan3793

**Comment by pan3793** [comment]
> > META-INF/services/jakarta.validation.valueextraction.ValueExtractor
> > META-INF/services/javax.validation.valueextraction.ValueExtractor
> 
> Looks like service loader interfaces

I think this is leaked. If you take a look at Jackson service loader interface, they should be relocated by ServicesResourceTransformer

**Comment by turboFei** [comment]
thanks, will check it

**Comment by turboFei** [comment]
Looks fine now.

for mvn:
```
jar -tf openapi/openapi-client/target/celeborn-openapi-client_2.12-0.6.0-SNAPSHOT.jar|grep ValueExtractor
org/apache/celeborn/shaded/org/openapitools/jackson/nullable/JsonNullableJakartaValueExtractor.class
org/apache/celeborn/shaded/org/openapitools/jackson/nullable/JsonNullableValueExtractor.class
org/apache/celeborn/shaded/org/openapitools/jackson/nullable/JsonNullableValueExtractorHelper$ValueSetter.class
org/apache/celeborn/shaded/org/openapitools/jackson/nullable/JsonNullableValueExtractorHelper.class
META-INF/services/org.apache.celeborn.shaded.jakarta.validation.valueextraction.ValueExtractor
META-INF/services/org.apache.celeborn.shaded.javax.validation.valueextraction.ValueExtractor

```

sbt:
```
jar -tf openapi/openapi-client/target/scala-2.12/celeborn-openapi-client_2.12-0.6.0-SNAPSHOT.jar|grep ValueExtractor
META-INF/services/jakarta.validation.valueextraction.ValueExtractor
META-INF/services/javax.validation.valueextraction.ValueExtractor
org/apache/celeborn/shaded/org/openapitools/jackson/nullable/JsonNullableJakartaValueExtractor.class
org/apache/celeborn/shaded/org/openapitools/jackson/nullable/JsonNullableValueExtractor.class
org/apache/celeborn/shaded/org/openapitools/jackson/nullable/JsonNullableValueExtractorHelper$ValueSetter.class
org/apache/celeborn/shaded/org/openapitools/jackson/nullable/JsonNullableValueExtractorHelper.class

```

**Comment by turboFei** [comment]
checking for SBT

File: project/CelebornBuild.scala [hunk 3]
```diff
@@ -1342,7 +1343,55 @@ object CelebornOpenApi {
...
+
+      (assembly / assemblyShadeRules) := Seq(
+        ShadeRule.rename("io.swagger.**" -> "org.apache.celeborn.shaded.io.swagger.@1").inAll,
+        ShadeRule.rename("org.openapitools.**" -> "org.apache.celeborn.shaded.org.openapitools.@1").inAll,
+        ShadeRule.rename("com.google.**" -> "org.apache.celeborn.shaded.com.google.@1").inAll,
+        ShadeRule.rename("javax.annotation.**" -> "org.apache.celeborn.shaded.javax.annotation.@1").inAll,
+        ShadeRule.rename("org.glassfish.jersey.**" -> "org.apache.celeborn.shaded.org.glassfish.jersey.@1").inAll,
+        ShadeRule.rename("com.fasterxml.jackson.**" -> "org.apache.celeborn.shaded.com.fasterxml.jackson.@1").inAll,
+        ShadeRule.rename("jakarta.validation.**" -> "org.apache.celeborn.shaded.jakarta.validation.@1").inAll,
+        ShadeRule.rename("javax.validation.**" -> "org.apache.celeborn.shaded.javax.validation.@1").inAll
```
Comment (turboFei) [inline project/CelebornBuild.scala:1384]: Fixed for sbt, @pan3793

File: openapi/openapi-client/pom.xml [hunk 4]
```diff
@@ -192,6 +192,78 @@
...
+              <pattern>com.fasterxml.jackson</pattern>
+              <shadedPattern>${shading.prefix}.com.fasterxml.jackson</shadedPattern>
+            </relocation>
+            <relocation>
+              <pattern>jakarta.validation</pattern>
+              <shadedPattern>${shading.prefix}.jakarta.validation</shadedPattern>
+            </relocation>
+            <relocation>
+              <pattern>javax.validation</pattern>
+              <shadedPattern>${shading.prefix}.javax.validation</shadedPattern>
```
Comment (turboFei) [inline openapi/openapi-client/pom.xml:230]: fixed for maven

**Review by pan3793 — APPROVED** [review]

**Pull request closed without merging**